src/parquet/column/scanner-test.cc

```diff
@@ -0,0 +1,386 @@
+ public:
+  void SetUp() {
+    values_.resize(0);
+    std::random_device rd;
+    SEED = rd();
+  }
+
+  void TearDown() {}
+
+  typedef typename type_traits<TYPE>::value_type T;
+
+  void InitScanner(Repetition::type rep,
+      int16_t max_def_level = 0, int16_t max_rep_level = 0) {
+    NodePtr type = schema::PrimitiveNode::Make("c1", rep, static_cast<Type::type>(TYPE));
+    descr_.reset(new ColumnDescriptor(type, max_def_level, max_rep_level));
+
+    std::unique_ptr<PageReader> pager(new test::MockPageReader(pages_));
+    scanner_ = Scanner::Make(ColumnReader::Make(descr_.get(), std::move(pager)));
+  }
+
+ protected:
+  void AddPages(size_t pages, size_t num_values,
+      std::vector<int16_t> def_levels, int16_t max_def_level,
+      std::vector<int16_t> rep_levels, int16_t max_rep_level) {
+    size_t page_size = std::max(def_levels.size(), num_values);
+    std::vector<T> page(num_values);
+    for (size_t i = 0; i < pages; i++) {
+      random_numbers(num_values, SEED, &page[0]);
+      buffers.push_back(std::vector<uint8_t>());
+      pages_.push_back(
+          MakeDataPage<TYPE>(page, def_levels, max_def_level,
+              rep_levels, max_rep_level, &buffers.back()));
+      values_.insert(values_.end(), page.begin(), page.end());
+    }
+  }
+
+  void VerifyOutput(size_t total_values) {
+    TypedScanner<TYPE>* scanner = reinterpret_cast<TypedScanner<TYPE>* >(scanner_.get());
+    T val;
+    bool is_null;
+    size_t j = 0;
+    for (size_t i = 0; i < total_values; i++) {
+      ASSERT_TRUE(scanner->HasNext());
+      ASSERT_TRUE(scanner->NextValue(&val, &is_null));
+      if (!is_null)
+        ASSERT_EQ(values_[j++], val);
+    }
+    ASSERT_FALSE(scanner->HasNext());
+  }
+
+  std::list< std::vector<uint8_t> > buffers;
+  std::unique_ptr<ColumnDescriptor> descr_;
+  std::vector< std::shared_ptr<Page> > pages_;
+  std::shared_ptr<Scanner> scanner_;
+  std::vector<T> values_;
+  std::vector<uint8_t> data_buffer_; // For BA and FLBA
+
+  uint32_t SEED;
+  uint32_t PAGES = 4;
+};
+
+template <>
+void TestFlatScanner<Type::BOOLEAN>::AddPages(size_t pages, size_t num_values,
+    std::vector<int16_t> def_levels, int16_t max_def_level,
+    std::vector<int16_t> rep_levels, int16_t max_rep_level) {
+  size_t page_size = std::max(def_levels.size(), num_values);
+  std::vector<bool> page;
+  for (size_t i = 0; i < pages; i++) {
+    page = flip_coins(num_values, 0.5);
+    buffers.push_back(std::vector<uint8_t>());
+    pages_.push_back(
+        MakeDataPage<Type::BOOLEAN>(page, def_levels, max_def_level,
+            rep_levels, max_rep_level, &buffers.back()));
+    values_.insert(values_.end(), page.begin(), page.end());
+  }
+}
+
+template <>
+void TestFlatScanner<Type::BYTE_ARRAY>::AddPages(size_t pages, size_t num_values,
+    std::vector<int16_t> def_levels, int16_t max_def_level,
+    std::vector<int16_t> rep_levels, int16_t max_rep_level) {
+  int max_byte_array_len = 12 + sizeof(uint32_t);
```

Comment:
This is fine, but I was confused by the presence of `sizeof(uint32_t)` -- shouldn't you pass `12` (in effect) as the last argument of `random_byte_array`? Otherwise you could possibly write past the end of `data_buffer_`.